Sources/SwiftSyntax/SourceLocation.swift

```diff
@@ -478,57 +423,283 @@ extension SyntaxProtocol {
+private struct SortedArray<Element>: RandomAccessCollection {
+  typealias Storage = [Element]
+  let storage: Storage
+
+  /// Initialize with the pre-sorted array.
+  ///
+  /// It is the caller's responsibility to pre-sort the array. Otherwise, the
+  /// behavior is undefined.
+  init(sortedArray: [Element]) {
+    self.storage = sortedArray
+  }
+
+  /// Find the first index where the element is ordered after `value`.
+  func upperBoundIndex<Value>(_ val: Value, by areIncreasingOrder: (Value, Element) -> Bool) -> Storage.Index {
+    var start = storage.startIndex
+    var end = storage.endIndex
+    while end > start {
+      let idx = (end - start) / 2 + start
+      if (!areIncreasingOrder(val, storage[idx])) {
+        start = idx + 1
+      } else {
+        end = idx
+      }
+    }
+    return start
+  }
+
+  /// Equivalent to `upperBoundIndex(val, by: <)`.
+  func upperBoundIndex(_ val: Element) -> Storage.Index where Element: Comparable {
+    self.upperBoundIndex(val, by: <)
+  }
+
+  var startIndex: Int { storage.startIndex }
+  var endIndex: Int { storage.endIndex }
+  subscript(index: Storage.Index) -> Element {
+    storage[index]
+  }
+  func makeIterator() -> Storage.Iterator {
+    storage.makeIterator()
+  }
+  func withContiguousStorageIfAvailable<R>(_ body: (UnsafeBufferPointer<Element>) throws -> R) rethrows -> R? {
+    try storage.withUnsafeBufferPointer(body)
+  }
+}
+
+extension SortedArray: Sendable where Element: Sendable {}
+
+/// Represent a virtual file region in the source file.
+@_spi(Compiler)
+public struct VirtualFile: Sendable {
+  /// Start position of the virtual file region
+  public var startPosition: AbsolutePosition
+  /// End position of the virtual file region
+  public var endPosition: AbsolutePosition
+  /// Virtual file name.
+  public var fileName: String
+  /// Line number offset from the physical line number.
+  public var lineOffset: Int
+  /// The position of the filename literal.
+  public var fileNamePosition: AbsolutePosition
+}
+
+@_spi(Compiler)
+public final class SourceLineTable: Sendable {
+  /// Array of the *end* of newline character of the each line.
+  private let lines: SortedArray<AbsolutePosition>
+
+  /// Array of virtual files declared in the source tree.
+  private let virtualFileTable: SortedArray<VirtualFile>
+
+  /// The end position of the file.
+  ///
+  /// If the last line was terminated with a newline character, this must be the
+  /// same value as `table.last`.
+  let endOfFile: AbsolutePosition
+
+  init(lines: [AbsolutePosition], virtualFiles: [VirtualFile], endOfFile: AbsolutePosition) {
+    precondition((lines.last ?? .startOfFile) <= endOfFile)
+    self.lines = SortedArray(sortedArray: lines)
+    self.virtualFileTable = SortedArray(sortedArray: virtualFiles)
+    self.endOfFile = endOfFile
+  }
+
+  /// Create an instance by traversing a Syntax tree.
+  public static func create(for tree: some SyntaxProtocol) -> SourceLineTable {
```

Comment:
Because I wanted to make `SourceLineTable` a `class` and class initializer doesn't allow delegating factory functions. I.e. `init(tree: Syntax) { self = computeLines(tree: tree) }` is not possible.

Reply:
Oh, interesting. I didn’t realize that. 👍🏽